OSDOCS-8804: Adding the MicroShift service to a blueprint section is … (#70554)
openshift/openshift-docs: modules/microshift-adding-service-to-blueprint.adoc

Adding the {microshift-short} RPM package to an Image Builder blueprint enables the build of a {op-system-ostree} image with {microshift-short} embedded.

* If your goal is a faster installation, use step 1 and create your own minimal blueprint file.

> * If your goal is a faster startup for {microshift-short}, use the generated blueprint in step 2. Using the generated blueprints is a longer installation process, but a faster start up because container references are accessed locally.

Reviewer: Faster installation and faster startup are by-products of having one or another blueprint type, but they are not goals.
I think we should rephrase this to emphasize that step 1 results a minimal microshift installation, while the step 2  includes all the RPM packages and container images. We can still mention that the former results in a faster installation, and the latter in faster startup.

+
[IMPORTANT]
====
Replace <minimal-microshift> in the following procedures with the name of the TOML you are using.

> ====

Reviewer: Can we remove this please and use a file name in the commands rather than `<some name>`?
The latter really annoys when copy-pasting commands for execution.

Reviewer: We have references to this `minimal-microshift` in a few places. I'd suggest to replace it by `microshift-blueprint` to be less ambiguous.

.Procedure

> . Use the following example to create your own <minimal-microshift> blueprint file:

Reviewer: ```suggestion
. Use the following example to create your own `microshift-blueprint.toml` blueprint file:
```

Adding the {microshift-short} RPM package to an Image Builder blueprint enables the build of a {op-system-ostree} image with {microshift-short} embedded.

> * Follow step 1 to create your own minimal blueprint file which results in a faster minimal {microshift-short} installation.
> * Follow Step 2 to use the generated blueprint for installation which includes all the RPM packages and container images. This is a longer installation process, but a faster start up because container references are accessed locally.

Reviewer 2: ```suggestion
To achieve a faster minimal {microshift-short} installation, create your own minimal blueprint file. This results in a faster installation process. After you have your blueprint file, use it for the installation. It contains all the required RPM packages and container images and enables a faster start as the container references are accessed locally. With this approach, you can optimize the installation process and achieve faster results.
```

+

> [IMPORTANT]
> ====
> Replace <microshift-blueprint> in the following procedures with the name of the TOML file you are using.
> ====

Reviewer 2: ```suggestion

```

Reviewer 2: We are already adding callout to highlight this.

.Procedure

> . Use the following example to create your own `microshift-blueprint.toml` blueprint file:

Reviewer 2: ```suggestion
. Create your own `microshift-blueprint.toml` blueprint file:
```

+

> .Custom Image Builder blueprint example

Reviewer 2: ```suggestion
.Example custom image builder blueprint file:
```

Adding the {microshift-short} RPM package to an Image Builder blueprint enables the build of a {op-system-ostree} image with {microshift-short} embedded.

> * Follow step 1 to create your own minimal blueprint file which results in a faster minimal {microshift-short} installation.

Reviewer 3: ```suggestion
* Follow step 1 to create your own minimal blueprint file which results in a faster {microshift-short} installation.
```

> * Follow Step 2 to use the generated blueprint for installation which includes all the RPM packages and container images. This is a longer installation process, but a faster start up because container references are accessed locally.

Reviewer 3: ```suggestion
* Follow step 2 to use the generated blueprint for installation which includes all the RPM packages and container images. This is a longer installation process, but a faster start up because container references are accessed locally.
```

+
[IMPORTANT]
====
* Replace _<microshift_blueprint.toml>_ in the following procedures with the name of the TOML file you are using.
* Replace _<microshift_blueprint>_ in the following procedures with the name you want to use for your blueprint.
====

.Procedure

. Use the following example to create your own blueprint file:
+
.Custom Image Builder blueprint example
+
[source,text]
----

> cat > microshift_blueprint.toml <<EOF <1>

Reviewer 3: ```suggestion
cat > _<microshift_blueprint.toml>_ <<EOF <1>
```

Author: This is an actual image builder blueprint example. so can this have <>?

> name = "microshift_blueprint" <2>

Reviewer 3: ```suggestion
name = "_<microshift_blueprint>_" <2>
```

Author: same comment as above

Reviewer 3: same reply

----
name = "microshift-4.15.1"
description = "MicroShift 4.15.1 on x86_64 platform"
version = "0.0.1"
modules = []
groups = []

[[packages]] <1>
name = "microshift"
version = "4.15.1"
...
...

[customizations.services] <2>
enabled = ["microshift"]

[customizations.firewall]
ports = ["22:tcp", "80:tcp", "443:tcp", "5353:udp", "6443:tcp", "30000-32767:tcp", "30000-32767:udp"]
...
...

[[containers]] <3>
source = "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:f41e79c17e8b41f1b0a5a32c3e2dd7cd15b8274554d3f1ba12b2598a347475f4"

[[containers]]
source = "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:dbc65f1fba7d92b36cf7514cd130fe83a9bd211005ddb23a8dc479e0eea645fd"
...
…
EOF
----
<1> References for all non-optional {microshift-short} RPM packages using the same version compatible with the `microshift-release-info` RPM.
<2> References for automatically enabling {microshift-short} on system startup and applying default networking settings.
<3> References for all non-optional {microshift-short} container images necessary for an offline MicroShift mode of operation.

. Add the blueprint to the Image Builder by running the following command:
+
[source,terminal]
----

> $ sudo composer-cli blueprints push <minimal-microshift>.toml <1>

Reviewer: ```suggestion
$ sudo composer-cli blueprints push microshift-blueprint.toml <1>
```

----

> <1> Replace <minimal-microshift> with the name of your TOML file as needed.

Reviewer: ```suggestion
```

.Verification

. Verify the Image Builder configuration listing only {microshift-short} packages by running the following command:
+
[source,terminal]
----

> $ sudo composer-cli blueprints depsolve <minimal-microshift> | grep microshift <1>

Reviewer: ```suggestion
$ sudo composer-cli blueprints depsolve microshift-blueprint | grep microshift
```

----

> <1> Replace <minimal-microshift> with the name of your TOML file as needed.

Reviewer: ```suggestion
```

The wildcard `*` in the commands uses the latest {microshift-short} RPMs. If you need a specific version, substitute the wildcard for the version you want. For example, insert `4.15.0` to download the {microshift-short} 4.15.0 RPMs.
====

> . Optional. Use the blueprint installed in the `/usr/share/microshift/blueprint` directory that is specific to platform architecture. See the following example snippet for an explanation of the blueprint sections:

Reviewer 3: ```suggestion
. Optional. Use the blueprint installed in the `/usr/share/microshift/blueprint` directory that is specific to your platform architecture. See the following example snippet for an explanation of the blueprint sections:
```

> . Optional. Use the blueprint installed in the `/usr/share/microshift/blueprint` directory that is specific to your platform architecture. See the following example snippet for an explanation of the blueprint sections:

Reviewer 2: ```suggestion
. Optional: Use the blueprint installed in the `/usr/share/microshift/blueprint` directory that is specific to your platform architecture:
```

+

> .Generated Image Builder blueprint example snippet

Reviewer 2: ```suggestion
.Example image builder blueprint file
```

+
[source,text]
----
name = "microshift-blueprint"
description = "MicroShift 4.15.1 on x86_64 platform"
version = "0.0.1"
modules = []
groups = []

[[packages]] <1>
name = "microshift"
version = "4.15.1"
...
...

[customizations.services] <2>
enabled = ["microshift"]

[customizations.firewall]
ports = ["22:tcp", "80:tcp", "443:tcp", "5353:udp", "6443:tcp", "30000-32767:tcp", "30000-32767:udp"]
...
...

[[containers]] <3>
source = "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:f41e79c17e8b41f1b0a5a32c3e2dd7cd15b8274554d3f1ba12b2598a347475f4"

[[containers]]
source = "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:dbc65f1fba7d92b36cf7514cd130fe83a9bd211005ddb23a8dc479e0eea645fd"
...
…
EOF
----

> <1> References for all non-optional {microshift-short} RPM packages using the same version compatible with the `microshift-release-info` RPM.
> <2> References for automatically enabling {microshift-short} on system startup and applying default networking settings.
> <3> References for all non-optional {microshift-short} container images necessary for an offline deployment.

Reviewer 2: ```suggestion
<1> References for all non-optional {microshift-short} RPM packages that use the same version compatible with the `microshift-release-info` RPM.
<2> References for the automatic activation of {microshift-short} on system startup and the application of default network settings.
<3> References for all non-optional {microshift-short} container images required for offline use.
```

> <3> References for all non-optional {microshift-short} container images necessary for an offline MicroShift mode of operation.

Reviewer 3: ```suggestion
<3> References for all non-optional {microshift-short} container images necessary for an offline deployment.
```

. Add the blueprint to the Image Builder by running the following command:
+
[source,terminal]
----

> $ sudo composer-cli blueprints push microshift-blueprint.toml <1>

Reviewer 3: ```suggestion
$ sudo composer-cli blueprints push <microshift-blueprint.toml> <1>
```
User-replaced values always have the carrots wherever they are.

----

> <1> Replace <microshift-blueprint.toml> with the name of your TOML file.

Reviewer 2: ```suggestion
<1> Specify the name of your TOML file.
``

Reviewer 2: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#callouts-in-code-blocks:~:text=%5Bsource%2Cterminal%5D%0A%2D%2D%2D%2D%0A%24%20oc%20autoscale,percent%20of%20requested%20CPU.

Reviewer 2: same suggestion for all instances.

> <1> Replace <microshift-blueprint> with the name of your TOML file as needed.

Reviewer: Do we need this repeated? We have it on the top of the doc.

Reviewer 3: @ggiguash Yes, that's docs style. We don't know if users are going to come straight to the procedure and not see the top note.

> <1> Replace <microshift-blueprint> with the name of your TOML file as needed.

Reviewer 3: ```suggestion
<1> Replace <microshift-blueprint.toml> with the name of your TOML file as needed.
```

.Verification

. Verify the Image Builder configuration listing only {microshift-short} packages by running the following command:
+
[source,terminal]
----

> $ sudo composer-cli blueprints depsolve <microshift-blueprint> | grep microshift <1>

Reviewer: I see that you added a template to the command: `<microshift-blueprint>`.
So, the <1> comment is not correct. It should be the `name` definition inside the TOML file.

Author: I can replace the existing reference with the below
<1> Replace <microshift-blueprint> with the name tag in your TOML file.

> $ sudo composer-cli blueprints depsolve microshift-blueprint | grep microshift <1>

Reviewer 3: ```suggestion
$ sudo composer-cli blueprints depsolve <microshift-blueprint> | grep microshift <1>
```

----

> <1> Replace <microshift-blueprint> with the name of your TOML file as needed.

Reviewer: Do we need this repeated? We have it on the top of the doc.

Reviewer 3: @ggiguash Yes, that's docs style. We don't know if users are going to come straight to the procedure and not see the top note.

> <1> Replace <microshift-blueprint> with the name of your TOML file as needed.

Reviewer 3: ```suggestion
<1> Replace <microshift-blueprint> with the name of your TOML file.
```

+
.Example output
+
[source,terminal]
----

> blueprint: microshift-blueprint v0.0.1

Reviewer 4: I know that the example is potentially outside the scope of the review, but shouldn't this reflect the version we are targetting? 4.13 seems out of place here.

Reviewer: Right, we can replace `4.13.1` by `4.15.1` at least.
I'm not sure what to do with the dates and commits though, i.e. `202305250827.p0.g4105d3`.

Maybe, we should remove this output all together?

Author: So Should I remove this output file?

+
[source,terminal]
----

> $ sudo composer-cli blueprints depsolve <minimal-microshift> <1>

Reviewer: ```suggestion
$ sudo composer-cli blueprints depsolve microshift-blueprint.toml <1>
```

> $ sudo composer-cli blueprints depsolve microshift-blueprint <1>

Reviewer 3: ```suggestion
$ sudo composer-cli blueprints depsolve <microshift-blueprint> <1>
```

> $ sudo composer-cli blueprints depsolve <microshift-blueprint> <1>

Author: I can replace the existing reference with the below
<1> Replace <microshift-blueprint> with the name tag in your TOML file.

> $ sudo composer-cli blueprints depsolve microshift-blueprint.toml <1>

Reviewer: ```suggestion
$ sudo composer-cli blueprints depsolve microshift-blueprint <1>
```

----

> <1> Replace <minimal-microshift> with the name of your TOML file as needed.

Reviewer: ```suggestion
```

> <1> Replace <microshift-blueprint> with the name of your TOML file as needed.

Reviewer: Do we need this repeated? We have it on the top of the doc.
